README
To enable it:
 * add `customPlugins = Lemonldap::NG::Portal::Plugins::DebianSSOCerts` in
   /etc/lemonldap-ng.ini, section [portal]
 * add a menu entry that points to `https://<portal>/certs`

Configuration:
 * to avoid mail filtering, change default "From" parameter in LLNG SMTP
   configuration

TODO:
 * improve CSS
 * improve all openssl commands (plugins directory)


debian/changelog
debian-sso (0.0.1) UNRELEASED; urgency=medium

  * Initial release

 -- Xavier Guimard <yadd@debian.org>  Tue, 24 Mar 2020 13:18:30 +0100


plugins/DebianUsers.pm
# Lemonldap::NG plugin to register authenticated users (GitHub, LinkedIn,...)
# but not known in our SQL database
#
# It handles 2 endpoints:
#  * LLNG internal getUser() method is replaced by check() which displays
#    form if needed
#  * https://portal/debianregister : receive registration data from form

package Lemonldap::NG::Portal::Plugins::DebianUsers;

use strict;
use Moo;

# Inheritance:
#  * LLNG plugin
#  * SQL object (uses DBI parameters declared in authentication scheme) - non-DD database
#    Provides $self->dbh, a DBI object
#  * LDAP object (uses LDAP parameters declared in authentication scheme) - DD database
#    Provides $self->ldap, a Net::LDAP object
extends 'Lemonldap::NG::Portal::Main::Plugin',
  'Lemonldap::NG::Portal::Lib::DBI',
  'Lemonldap::NG::Portal::Lib::LDAP';

# LLNG constants used in `check()` responses
use Lemonldap::NG::Portal::Main::Constants qw(
  PE_OK
  PE_INFO
  PE_SENDRESPONSE
);

# Parameters from lemonldap-ng.ini with their default values
#
# Timeout to register, default 10 mn
has debianRegisterTimeout => (
    is      => 'ro',
    default => sub {
        $_[0]->conf->{debianRegisterTimeout} || 600;
    },
);

# Nickname regexp, default 3 to 20 alphanum-or-"-" chars. Starts with a
# alphanum char
has debianNicknameRegexp => (
    is      => 'ro',
    default => sub {
        my $re = $_[0]->conf->{debianNicknameRegexp} || '^\w[\w\-]{2,19}$';
        return qr/$re/;
    },
);

# SQL fields
# Required:
# * uid: LinkedIn uid
# * mail: mail given by LinkedIn (may be changed later)
# * username: chosen Debian uid
# * displayname: fullname to display
# * firstname
# Optional:
# * lastname
# * gpgkey: GPG key
# * sshkey: SSK key
our @fields =
  (qw(uid mail username displayname firstname lastname gpgkey sshkey));

# NB: uid and mail are required but provided by authentication mechanism
our @requiredFields = (qw(username displayname firstname));

# Sessions parameter: this plugin uses macros that returns wanted data:
#  * mail: should be
#    `($linkedIn_emailAddress ? $linkedIn_emailAddress : $github_emailAddress ? $github_emailAddress : "$uid\@debian.org" )`

# Internal getters
#
# One-time-token instance
has ott => (
    is      => 'ro',
    lazy    => 1,
    default => sub {
        my $ott = $_[0]->{p}->loadModule('::Lib::OneTimeToken');
        $ott->timeout( $_[0]->debianRegisterTimeout );
        return $ott;
    }
);

# INTERFACE WITH LLNG
#
# LLNG Entry point to catch authentication requests before "search" step:
# launches check() instead of LLNG internal getUser() method
use constant aroundSub => { getUser => 'check' };

# Declare /debianregister route (used by registration form).
# Only POST requests are accepted here
#
# No need to launch  ::Lib::DBI and ::Lib::LDAP init() methods here: already
# launched by authentication modules
sub init {
    my ($self) = @_;
    $self->addUnauthRoute( debianregister => 'register', ['POST'] );
    # Tests
    my $res = $self->Lemonldap::NG::Portal::Lib::DBI::init();
    return $res if($res);
    unless($self->table) {
        $self->error('Set SQL table in configuration');
        return 0;
    }
    return 1;
}

# RUNNING METHODS

# Checks if non-DD user is known from SQL database. Displays registration
# form else
#
# This method is launched during auth process, it must returns LLNG constants
#
# Response: Lemonldap::NG::Portal::Main::Constants constant
sub check {
    my ( $self, $sub, $req ) = @_;

    # Don't intercept DD requests
    return $sub->($req) if $self->p->getModule( $req, "auth" ) eq 'LDAP';

    # Get linkedIn/GitHub data
    $self->p->setAuthSessionInfo($req);

    # Map social network id to user field
    $req->user($req->sessionInfo->{linkedIn_id} || $req->sessionInfo->{github_id});

    # Launch getUser and intercept result
    my $res = $sub->($req);

    # If getUser succeeds, no need to do anything
    return PE_OK if $res == PE_OK;

    # USER NOT FOUND IN USERS DATABASE

    # Calculate macros
    $self->p->setMacros($req);

   # Display registration form: since we are in auth process, direct PSGI output
   # is not allowed here, we must use this hook
    $req->response( $self->form($req) );
    return PE_SENDRESPONSE;
}

# Register user
#
# This method receives POST requests from registration form (POST requests only
# as declared in init() method)
#
# Response: PSGI format
sub register {
    my ( $self, $req ) = @_;

    # Objects inherited from LLNG libs (configured via manager:
    # "General Parameters" => "authentication modules"):
    #  * $self->dbh : SQL  connection (non DD users, registered using this
    #                 plugin)
    #  * $self->ldap: LDAP connection (DD users)

    # 0 - get data
    my $nickname = $req->param('username');
    my $token    = $req->param('token');

    # 1 - check token
    my $session = $self->ott->getToken($token);
    unless ($session) {
        return $self->sendError( $req,
            'Bad token or token expired, restart registration process', 400 );
    }
    $req->user( $session->{user} );
    $req->sessionInfo( $session->{data} );

    # Check username string
    if ( $nickname !~ $self->debianNicknameRegexp ) {
        return $self->form( $req, 'Nickname not allowed' );
    }

    # Check other fields
    my $error = $self->_checkOtherFields($req);
    if ($error) {
        return $self->form( $req, $error );
    }

    # 2 - verify that nickname does not already exist in LDAP or SQL DB
    my $res = eval { $self->_checknickname($nickname) };
    if ($@) {
        return $self->p->sendError( $req, 'Registration databases seems down',
            500 );
    }

    # 2.1 - case OK (nickname not used)
    if ($res) {
        return $self->_registerUser($req);
    }

    # 2.2 - case NOK, display form
    else {
        return $self->form( $req,
            'Nickname already exists, choose another one' );
    }
}

# Internal method to populate and display form
#
# Response: PSGI format
sub form {
    my ( $self, $req, $error ) = @_;
    my $token = $self->ott->createToken(
        { user => $req->user, data => $req->sessionInfo } );
    if ( $req->wantJSON ) {
        return $self->p->sendJSONresponse( $req,
            { result => ( $error ? 0 : 1 ), error => $error, token => $token, }
        );
    }
    else {
        return $self->p->sendHtml(
            $req,
            'debianRegistration',
            params => {
                USER  => $req->user,
                TOKEN => $token,
                ERROR => $error,
            }
        );
    }
}

# Internal boolean function that checks if proposed nickname is already used
sub _checknickname {
    my ( $self, $nickname ) = @_;

    # 1 - LDAP
    $self->validateLdap;

    unless ( $self->ldap ) {
        $self->logger->error('LDAP seems down');
        die;
    }

    $self->bind();

    my $mesg = $self->ldap->search(
        base   => $self->conf->{ldapBase},
        scope  => 'sub',
        filter => $self->filter->( { user => $nickname } ),
        defer  => $self->conf->{ldapSearchDeref} || 'find',
        attrs  => $self->attrs,
    );
    if ( $mesg->code() != 0 ) {
        $self->logger->error( 'LDAP returned an error: ' . $mesg->error );
        die;
    }

    # Fail if a DD exists with this nickname
    return 0 if $mesg->entry(0);

    # 2 - SQL DB
    my $sth;
    eval {
        $sth = $self->dbh->prepare(
            'SELECT username from ' . $self->table . ' WHERE username=?' );
        $sth->execute($nickname);
    };
    if ($@) {
        $self->logger->error("SQL DB returned an error: $@");
        die;
    }
    return 0 if $sth->fetchrow_hashref();

    # Well, nickname not found anywhere, let's agree this nickname
    return 1;
}

# Internal function that checks if fields are well filed.
# Return an error to display if needed
sub _checkOtherFields {
    my ( $self, $req ) = @_;

    # Check if required fields are set
    my @missings;
    foreach (@requiredFields) {
        push @missings, $_ unless $req->param($_);
    }
    if (@missings) {
        return 'These fields are required: ' . join( ', ', @missings );
    }

    # TODO: insert here fields checks

    # No error:
    return '';
}

# Real registration method: insert fields in SQL DB
#
# Response: PSGI format
sub _registerUser {
    my ( $self, $req ) = @_;
    my $sth;

    # TODO: fix that
    my %values = ( map { ( $_ => $req->param($_) ) } @fields );
    $values{uid}  = $req->sessionInfo->{linkedIn_id};
    $values{mail} = $req->sessionInfo->{linkedIn_emailAddress};

    # Filter fields: only those who are filed are taken in this query
    my @_fields = map { $values{$_} ? ($_) : () } @fields;
    eval {
        $sth =
          $self->dbh->prepare( 'INSERT INTO '
              . $self->table . ' ('
              . join( ',', @_fields )
              . ") VALUES ("
              . join( ",", map { '?' } @_fields )
              . ")" );
        $sth->execute( map { $values{$_} } @_fields );
    };
    if ($@) {

        # If connection isn't available, error is displayed by dbh()
        $self->logger->error("DBI error: $@") if ( $self->_dbh );
        return $self->form( $req, 'Error form database, try later' );
    }

    # Finish auth process and display registration message

    # Prepare message (HTML acccepted here)
    $req->info('Successfully registered');

    # Restart Choices initialization
    $self->p->extractFormInfo($req);
    $self->p->getUser($req) || $self->logger->error("User registered but not found, verify your configuration");;

    # Launch remaining LLNG methods (`do()` from
    # Lemonldap::NG::Portal::Main::Run will do the job. Just to declare
    # methods to launch (keeping authentication from remote system)
    return $self->p->do(
        $req,
        [

            # Plugins entryPoints
            @{ $self->p->betweenAuthAndData },

            # Methods that populates session data
            $self->p->sessionData,

            # Plugins entryPoints
            @{ $self->p->afterData },

            # Methods that validate session
            $self->p->validSession,

            # Plugins entryPoints
            @{ $self->p->endAuth },

            # Fake entrypoint to force displaying registration message
            sub { PE_INFO },
        ]
    );
}

1;


static/debianca/index.js
function populateAttr(cert, attr, value) {
    cert.addSubjectAttr(attr, value);
}

function pem2pkcs12(cert,key){
    let fCert = forge.pki.certificateFromPem(cert);
    let fKey = forge.pki.privateKeyFromPem(key);
    let password = prompt("Enter export password");
    let p12Asn1 = forge.pkcs12.toPkcs12Asn1(fKey, fCert, password,{algorithm:'3des'});
    // base64-encode p12
    let p12Der = forge.asn1.toDer(p12Asn1).getBytes();
    return p12Der;
}

function downloadP12B64(name, p12b64) {
    var a = document.createElement('a');
    a.setAttribute('href', 'data:application/x-pkcs12;base64,' + p12b64);
    a.setAttribute('download', name);
    //a.appendChild(doc if (document.createEvent) {

    if (document.createEvent) {
        var event = document.createEvent('MouseEvents');
        event.initEvent('click', true, true);
        a.dispatchEvent(event);
    }
    else {
        a.click();
    }
}

$(document).ready(function() {

  // CSR generated by browser

  $('#getcert').on('click', function() {
    console.debug('generate keys');
    $('#getcert').prop('disabled', true);
    let pkcs10CSR = new org.pkijs.simpl.PKCS10();

    let publicKey;
    let privateKey;

    let hash_algorithm = "sha-512";
    let signature_algorithm_name = "RSASSA-PKCS1-V1_5"; // "RSA-PSS", "ECDSA"

    // WebCrypto
    let crypto = org.pkijs.getCrypto();
    if (typeof crypto == "undefined") {
        alert("No WebCrypto extension found");
        return;
    }

    pkcs10CSR.version = 0;
    pkcs10CSR.attributes = [];
    populateAttr(pkcs10CSR,'CN',$('#mail').val());
    populateAttr(pkcs10CSR,'O','Debian');
    populateAttr(pkcs10CSR,'OU','Debian Developers');

    Promise.resolve().then(function () {
      let algorithm = org.pkijs.getAlgorithmParameters(signature_algorithm_name, "generatekey");
        if ("hash" in algorithm.algorithm) {
            algorithm.algorithm.hash.name = hash_algorithm;
        }
        return crypto.generateKey(algorithm.algorithm, true, algorithm.usages);
    }).then(function (keyPair) {
        publicKey = keyPair.publicKey;
        privateKey = keyPair.privateKey;
    }).catch(function (error) {
        alert("Error during key generation: " + error);
    }).then(function () {
        return pkcs10CSR.subjectPublicKeyInfo.importKey(publicKey);
    }).then(function (result) {
        return crypto.digest({name: "SHA-1"}, pkcs10CSR.subjectPublicKeyInfo.subjectPublicKey.value_block.value_hex);
    }).then(function (result) {
        pkcs10CSR.attributes.push(new org.pkijs.simpl.ATTRIBUTE({
            type: "1.2.840.113549.1.9.14", // pkcs-9-at-extensionRequest
            values: [(new org.pkijs.simpl.EXTENSIONS({
                extensions_array: [
                    new org.pkijs.simpl.EXTENSION({
                        extnID: "2.5.29.14",
                        critical: false,
                        extnValue: (new org.pkijs.asn1.OCTETSTRING({value_hex: result})).toBER(false)
                    })
                ]
            })).toSchema()]
        }));
    }).then(function () {// Signing final PKCS#10 request
        return pkcs10CSR.sign(privateKey, hash_algorithm);
    }).catch(function (error) {
        alert("Error during exporting public key: " + error);
    }).then(function () {
        return crypto.exportKey("pkcs8", privateKey);
    }).then(function (pkcs8Privkey) {
        $.ajax('/certs/enroll', {
          dataType: 'json',
          type: 'POST',
          data: {
            csr: csr2pem(pkcs10CSR),
            token: $('#token').val(),
            comment: $('#id_comment').val(),
            validity: $('#id_validity').val(),
          },
          success: function(data) {
            if(data.result == 1) {
              let pem = data.cert;
              let p12Der = pem2pkcs12(pem,privkey2pem(pkcs8Privkey));
              let p12b64 = forge.util.encode64(p12Der);
              $('#catitle').html('Certificate generated');

              downloadP12B64('client.p12',p12b64);
            }
            else {
              console.error('Bad response',data);
            }
          },
          error: function(err) {
            console.error('Error: ', err);
            $('#catitle').html('An error occured, this may be an expired token, try to reload this page');
            $('#catitle').removeClass('alert-success').addClass('alert-danger');
          },
        });
    }).catch(function (error) {
        alert("Error signing PKCS#10: " + error);
    });
    return false;
  });
})


plugins/DebianSSOCerts.pm
package Lemonldap::NG::Portal::Plugins::DebianSSOCerts;

use strict;

use Moo;
use IPC::Run;
use Date::Parse;
use File::Temp;

extends 'Lemonldap::NG::Portal::Main::Plugin',
  'Lemonldap::NG::Portal::Lib::SMTP';

# lemonldap-ng.ini parameters (section [portal]
has openssl => (
    is      => 'ro',
    default => sub {
        $_[0]->conf->{openssl} || '/var/lib/debian-sso/openssl.sh';
    },
);

has gpgDatabases => (
    is      => 'ro',
    default => sub {
        $_[0]->conf->{gpgDatabases}
          || '/usr/share/keyrings/debian-nonupload.gpg /usr/share/keyrings/debian-keyring.gpg';
    },
);

has gpgCertTokenTimeout => (
    is      => 'ro',
    default => sub {
        $_[0]->conf->{gpgCertTokenTimeout} || 600;
    },
);

has highCertAuthnLevel => (
    is      => 'ro',
    default => sub {
        $_[0]->conf->{highCertAuthnLevel} || 5;
    },
);

has mailAttribute => (
    is      => 'ro',
    default => sub {
        $_[0]->conf->{mailAttribute} || 'mail';
    },
);

has opensslSignArgs => (
    is      => 'ro',
    default => sub {
        $_[0]->conf->{opensslSignArgs}
          || 'x509 -req -CA ca.crt -CAkey ca.key -CAcreateserial -days';
    },
);

has opensslHighSignArgs => (
    is      => 'ro',
    lazy    => 1,
    default => sub {
        $_[0]->opensslSignArgs;
    },
);

has ott => (
    is      => 'rw',
    lazy    => 1,
    default => sub {
        my $ott = $_[0]->{p}->loadModule('::Lib::OneTimeToken');
        return $ott;
    }
);

has ottSecure => (
    is      => 'rw',
    lazy    => 1,
    default => sub {
        my $ott = $_[0]->{p}->loadModule('::Lib::OneTimeToken');
        $ott->timeout( $_[0]->gpgCertTokenTimeout );
        return $ott;
    }
);

has tmpDir => (
    is      => 'ro',
    default => sub {
        return File::Temp::tempdir( CLEANUP => 1 );
    },
);

sub init {
    my ($self) = @_;
    $self->addAuthRouteWithRedirect(
        certs => {
            enroll       => 'enroll',
            enrollSecure => 'enrollSecure',
            '*'          => 'run',
        },
        [ 'GET', 'POST' ]
    );
    $self->addAuthRouteWithRedirect(
        certs => { enroll => 'signCSR' },
        ['POST']
    );
    return 1;
}

# This method send a GPG encrypted message and upgrade authentication level
# then
sub enrollSecure {
    my ( $self, $req ) = @_;
    if ( my $token = $req->param('secureToken') ) {
        my $value;
        if (    $value = $self->ottSecure->getToken($token)
            and $value->{secureCert} )
        {
            $self->p->updateSession( $req,
                { authenticationLevel => $self->highCertAuthnLevel } );
        }
        else {
            return $self->p->sendError( $req, 'Token expired' );
        }
    }
    if ( $req->userData->{authenticationLevel} < $self->highCertAuthnLevel ) {
        my $mail = $req->userData->{ $self->mailAttribute }
          or return $self->p->sendError( $req, 'Unable to retrieve mail', 500 );

        my $token = $self->ottSecure->createToken( { secureCert => 1 } );
        my $url =
          $self->conf->{portal} . "certs/enrollSecure?secureToken=$token";
        my $txt = "Go to $url to get a higher authentication level";
        my ( $out, $err );
        IPC::Run::run( [
                qw(gpg --list-keys --homedir),
                $self->tmpDir,
                (
                    map { ( '--keyring', $_ ) } split /,?\s+/,
                    $self->gpgDatabases
                ),
                $mail
            ],
            ,
            \$out,
            \$err,
        );

        if ($?) {
            $self->userLogger->notice(
                "gpg failed to find user in database: $err");
            return $self->p->sendError( $req,
q{you requested this, but you can't use it unless your official key made it all the way into the package and found a release}
            );
        }
        IPC::Run::run( [
                qw(gpg --encrypt --armor --yes --trust-model always -r),
                $mail,
                (
                    map { ( '--keyring', $_ ) } split /,?\s+/,
                    $self->gpgDatabases
                ),
            ],
            \$txt,
            \$out,
            \$err,
        );
        if ($?) {
            $self->logger->error("gpg error: $err");
            return $self->p->sendError( $req, 'gpg failed' );
        }
        $self->send_mail( $mail, 'High verification mail', $out );
        return $self->p->sendHtml(
            $req, 'error',
            params => {
                RAW_ERROR       => 1,
                AUTH_ERROR_TYPE => 'warning',
                RAW_ERROR => 'GPG encrypted message sent, click on the link'
            }
        );
    }
    else {
        return $self->enroll($req);
    }
}

# Display cert list
sub run {
    my ( $self, $req ) = @_;
    my $certs = $self->getCurrentCerts($req);
    if ( $req->method eq 'POST' ) {

        # Revoke certificate
        my $serial = $req->param('serial');
        my $certid = $serial;
        $certid =~ s/://g;

        # Insert here CRL update and generation using $certid

        # Update cert list
        @$certs = map { $_->{serial} eq $serial ? () : $_ } @$certs;
        $self->p->updatePersistentSession( $req, { debianCerts => $certs } );
    }
    return $self->p->sendHtml( $req, 'certs', params => { CERTS => $certs } );
}

# Display CSR enrollment page
sub enroll {
    my ( $self, $req ) = @_;
    return $self->p->sendHtml(
        $req,
        'certenroll',
        params => {
            LEVEL => (
                $req->userData->{authenticationLevel} ==
                  $self->highCertAuthnLevel
            ),
            TOKEN => $self->ott->createToken(
                { id => $req->userData->{_session_id} }

                #{ user => $req->userData }
            ),
            MAIL => $req->userData->{ $self->mailAttribute },
        }
    );
}

# Sing CSR
sub signCSR {
    my ( $self, $req ) = @_;
    my $highLevel =
      ( $req->userData->{authenticationLevel} == $self->highCertAuthnLevel );
    my $id = $self->ott->getToken( $req->param('token') );
    unless ( $id and $id->{id} eq $req->userData->{_session_id} ) {
        $self->userLogger->notice('Bad token');
        return $self->p->sendError( $req, 'Bad token' );
    }
    my $csr  = $req->param('csr');
    my $days = $req->param('validity') || 365;
    $days = 365 if $days > 365 or $days < 1;
    my ( $out, $err );

    # Check CSR
    IPC::Run::run( [ $self->openssl, qw'req -noout -text' ],
        \$csr, \$out, \$err, );
    if ($?) {
        $self->userLogger->error("Bad CSR request: $err");
        $self->p->sendError( $req, 'Bad request' );
    }

    # Insert here request checks
    if (
        $out !~ /Subject: CN = (\S+) \+ O = Debian \+ OU = Debian Developers/s )
    {
        $self->userLogger->error("Bad CSR : $out");
        return $self->p->sendError( $req, 'Bad CSR request: ' . $out );
    }
    my $cn       = $1;
    my $userMail = $req->userData->{ $self->mailAttribute };
    if ( $cn !~ /\Q$userMail\E/ ) {
        return $self->p->sendError( $req,
            "Mail mistamtch in CN, this should be $userMail", 400 );
    }

    # Sign cert
    my $cmd =
      ( $highLevel ? $self->opensslHighSignArgs : $self->opensslSignArgs );
    IPC::Run::run( [ $self->openssl, ( split /\s+/, $cmd ), $days ],
        \$csr, \$out, \$err, );
    if ($?) {
        $self->userLogger->error("Bad CSR request: $err");
        $self->p->sendError( $req, 'Bad request' );
    }
    my $crt = $out;

    # Retrive serial number
    IPC::Run::run( [ $self->openssl, qw'x509 -noout -text' ],
        \$crt, \$out, \$err, );
    if ($?) {
        $self->logger->error("Unable to read generated certificate: $err");
        $self->p->sendError( $req, 'openssl error', 500 );
    }
    unless ( $out =~ /Serial Number:\s+([0-9a-f:]+)\n/s ) {
        $self->logger->error("Unable to find serial number in: $out");
        return $self->p->sendError( $req, 'Something wrong happens...', 500 );
    }
    my $serial = $1;
    $out =~ /Not After : (.*?)\n/;
    my $expires     = $1;
    my $currentCert = $self->getCurrentCerts($req);
    push @{$currentCert},
      {
        serial  => $serial,
        expires => $expires,
        comment => $req->param('comment') || '',
        ( $highLevel ? ( highLevel => 1 ) : () ),
      };
    $self->p->updatePersistentSession( $req, { debianCerts => $currentCert } );

    if ( $req->wantJSON ) {
        return $self->p->sendJSONresponse( $req,
            { result => 1, cert => $crt } );
    }
    else {
        return ( 200, [ 'Content-Type' => 'text/plain' ], [$crt] );
    }
}

# Internal methods

sub getCurrentCerts {
    my ( $self, $req ) = @_;
    my $user = $req->userData->{ $self->conf->{whatToTrace} };
    my $currentCert =
      $self->p->getPersistentSession($user)->data->{debianCerts} || [];
    my $now = time;
    foreach (@$currentCert) {
        print STDERR '# ' . str2time( $_->{expires} ) . " $now\n";
        $_->{expired} = 1 if str2time( $_->{expires} ) < $now;
    }
    return $currentCert;
}

1;
